pennylane/transforms/optimization/relative_phases.py

```diff
@@ -0,0 +1,234 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Transforms lowering gates and series of gates involving relative phases.
+Transformations first published in:
+
+Maslov, Dmitri. "On the Advantages of Using Relative Phase Toffolis with an Application to
+Multiple Control Toffoli Optimization", arXiv:1508.03273, arXiv, 2016.
+doi:10.48550/arXiv.1508.03273.
+
+Giles, Brett, and Peter Selinger. "Exact Synthesis of Multiqubit Clifford+T Circuits",
+arXiv:1212.0506, arXiv, 2013. doi:10.48550/arXiv.1212.0506.
+"""
+
+import pennylane as qml
+from pennylane.tape import QuantumScript, QuantumScriptBatch
+from pennylane.transforms import transform
+from pennylane.typing import PostprocessingFn
+
+from .pattern_matching import pattern_matching_optimization
+
+
+@transform
+def replace_relative_phase_toffoli(
+    tape: QuantumScript,
+) -> tuple[QuantumScriptBatch, PostprocessingFn]:
+    """Quantum transform to replace 4-qubit relative phase Toffoli gates, given in
```

Comment:
I would also mention that this uses pattern matching optimization with a given template.  So it's not strictly replacing the first component with the second. It's looking for subcircuits from the full pattern that can replaced by the rest of the pattern.

For example, this can cause confusion: https://github.com/PennyLaneAI/pennylane/issues/6135